### PR TITLE
Add Monitoring to GoForIt Client

### PR DESCRIPTION
**Pull request by asoni-stripe**
This PR is to add monitoring to GoForIt. The two checks are for:

- When was the backend last updated? For example, the JSON file has an "updated" flag to indicate this (presumably generated sensibly e.g: at time of database dump). If the backend hasn't been updated after some threshold (5 minutes), then push to DataDog. 

- When was the last time the GoForIt client refreshed the flags? It is possible the backend has been updated but GoForIt hasn't pulled in the changes in a while. We push to DataDog if this time exceeds our staleness threshold as well. 

Eventually we would like to send errors to Sentry as well, but leaving as a TODO for now. 

Using last modified time on the filesystem for the CSV file, but can remove if unwanted.

### CODE REVIEW

File: flags.go [hunk 1]
```diff
@@ -59,18 +66,48 @@ func (b csvFileBackend) Refresh() (map[string]Flag, error) {
...
+	defer f.Close()
+	statinfo, err := f.Stat()
+	if err != nil {
+		return time.Time{}, err
+	}
+	return statinfo.ModTime(), nil
+}
+
+// for json file, update time comes from json directly
+func (b jsonFileBackend) UpdateTime() (time.Time, error) {
```
Comment (vasi-stripe) [inline flags.go]: `UpdateTime` should be linked to the time the file was read. Can we maintain this in an member of the `jsonFileBackend` instance?

File: flags.go [hunk 2]
```diff
@@ -84,6 +121,17 @@ func Enabled(ctx context.Context, name string) (enabled bool) {
...
+	defer func() {
+		updateMtx.RLock()
+		staleness := time.Since(lastUpdate)
+		//histogram of cache process age
+		stats.Histogram("goforit.flags.process_cache_age_s", staleness.Seconds(), nil, .01)
+		if staleness > STALENESS_THRESHOLD && time.Since(lastAssert) > 60*time.Second {
+			lastAssert = time.Now()
+			//TODO send to sentry
+		}
+		updateMtx.RUnlock()
```
Comment (vasi-stripe) [inline flags.go]: Use defer for this. Yes, you can defer inside a defer :)

File: flags.go [hunk 3]
```diff
@@ -84,6 +121,17 @@ func Enabled(ctx context.Context, name string) (enabled bool) {
...
 	}()
 
+	defer func() {
+		updateMtx.RLock()
+		staleness := time.Since(lastUpdate)
+		//histogram of cache process age
+		stats.Histogram("goforit.flags.process_cache_age_s", staleness.Seconds(), nil, .01)
+		if staleness > STALENESS_THRESHOLD && time.Since(lastAssert) > 60*time.Second {
+			lastAssert = time.Now()
+			//TODO send to sentry
```
Comment (vasi-stripe) [inline flags.go]: For now, let's just log a message here

File: flags.go [hunk 4]
```diff
@@ -29,6 +31,11 @@ type Backend interface {
 	Refresh() (map[string]Flag, error)
 }
 
+type UpdatableBackend interface {
```
Comment (vasi-stripe) [inline flags.go]: We should document what this means :)  Also let's try and choose a better name, my initial stab was super confusing

File: flags.go [hunk 5]
```diff
@@ -178,6 +231,28 @@ func BackendFromJSONFile(filename string) Backend {
 // Consul key/value storage.
 func RefreshFlags(backend Backend) error {
 
+	updateableB, ok := backend.(UpdatableBackend)
```
Comment (vasi-stripe) [inline flags.go]: We should do this after the refresh

File: flags.go [hunk 6]
```diff
@@ -178,6 +231,28 @@ func BackendFromJSONFile(filename string) Backend {
...
+	if ok {
+		t, err := updateableB.UpdateTime()
+		if err == nil {
```
Comment (vasi-stripe) [inline flags.go]: We should handle the error case!

File: flags.go [hunk 7]
```diff
@@ -19,6 +19,8 @@ const statsdAddress = "127.0.0.1:8200"
 
 var stats *statsd.Client
 
+var STALENESS_THRESHOLD time.Duration = 5 * time.Minute
```
[the remark this answers is not in the document]
Comment (vasi-stripe) [reply flags.go]: Or actually, maybe we should make this configurable!

File: flags.go [hunk 8]
```diff
@@ -178,6 +231,28 @@ func BackendFromJSONFile(filename string) Backend {
...
+			staleness := time.Since(t)
+			stale := staleness > STALENESS_THRESHOLD
+			//histogram of staleness
+			stats.Histogram("goforit.flags.cache_file_age_s", staleness.Seconds(), nil, .1)
+			//gauge
+			var gauge float64
+			if stale {
+				gauge = 1
+			}
+			stats.Gauge("goforit.flags.cache_file_stale", gauge, nil, 1)
```
Comment (vasi-stripe) [inline flags.go]: Let's just log a message for staleness here, same as above.

File: flags.go [hunk 9]
```diff
@@ -59,18 +66,48 @@ func (b csvFileBackend) Refresh() (map[string]Flag, error) {
...
-	Flags []Flag `json:"flags"`
+	Flags       []Flag  `json:"flags"`
+	UpdatedTime float64 `json:"updated"`
 }
 
 var flags = map[string]Flag{}
 var flagsMtx = sync.RWMutex{}
 
+var lastUpdate time.Time
+var updateMtx = sync.RWMutex{}
```
Comment (vasi-stripe) [inline flags.go]: I feel like it's ok to use the same mutex for both updates and flags.

**Comment by vasi-stripe** [comment]
Looks good overall! A bunch of little changes needed.

File: flags.go [hunk 10]
```diff
@@ -84,6 +121,17 @@ func Enabled(ctx context.Context, name string) (enabled bool) {
 		stats.Gauge("goforit.flags.enabled", gauge, []string{fmt.Sprintf("flag:%s", name)}, .1)
 	}()
 
+	defer func() {
+		updateMtx.RLock()
+		staleness := time.Since(lastUpdate)
```
Comment (vasi-stripe) [inline flags.go]: What happens if we're not using an UpdatableBackend?

Comment (asoni-stripe) [reply flags.go]: This is separate from the backend being Updatable. lastUpdate tracks the last time the goforit client ran the refreshflags cycle. I can see how the naming is confusing here, I will rename this var.

File: flags.go (hunk 2 above)
Comment (asoni-stripe) [reply flags.go]: ha! i was wondering but it seemed like a crazy thought!

File: flags.go (hunk 4 above)
Comment (asoni-stripe) [reply flags.go]: Removed this interface...tacked on "age" to the existing Refresh method..let me know if better naming is needed there as well!

File: flags.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (asoni-stripe) [reply flags.go]: Done!

File: flags.go [hunk 11]
```diff
@@ -59,18 +66,48 @@ func (b csvFileBackend) Refresh() (map[string]Flag, error) {
 	return readFile(b.filename, "csv", parseFlagsCSV)
 }
 
+// for csv file, update time looks at last modified time on filesystem
```
[the remark this answers is not in the document]
Comment (asoni-stripe) [reply flags.go]: Removed

File: flags.go (hunk 1 above)
Comment (asoni-stripe) [reply flags.go]: Done..."age" is now returned in the Refresh() method, at same time the file was read

File: flags.go (hunk 9 above)
Comment (asoni-stripe) [reply flags.go]: Done

File: flags.go (hunk 3 above)
Comment (asoni-stripe) [reply flags.go]: Added some logging throughout the file, but just does a Printf for now

File: flags.go (hunk 5 above)
Comment (asoni-stripe) [reply flags.go]: Done

File: flags.go (hunk 6 above)
Comment (asoni-stripe) [reply flags.go]: Done

File: flags.go (hunk 8 above)
Comment (asoni-stripe) [reply flags.go]: Have added logging

File: flags.go [hunk 12]
```diff
@@ -19,14 +20,17 @@ const statsdAddress = "127.0.0.1:8200"
 
 var stats *statsd.Client
 
+var staleness_threshold time.Duration = 5 * time.Minute
+var staleMtx = sync.RWMutex{}
```
Comment (vasi-stripe) [inline flags.go]: I think we can just use one big mutex for all globals. The more mutexes we have, the harder it gets to reason about what's going on.

File: flags.go [hunk 13]
```diff
@@ -37,25 +41,26 @@ type jsonFileBackend struct {
...
+func readFile(file string, backend string, parse func(io.Reader) (map[string]Flag, time.Time, error)) (map[string]Flag, time.Time, error) {
 	var checkStatus statsd.ServiceCheckStatus
 	defer func() {
 		stats.SimpleServiceCheck("goforit.refreshFlags."+backend+"FileBackend.present", checkStatus)
 	}()
 	f, err := os.Open(file)
 	if err != nil {
 		checkStatus = statsd.Warn
-		return nil, err
+		log.Print("unable to open backend file:: ", err)
```
Comment (vasi-stripe) [inline flags.go]: This looks a bit odd. There should be just one colon, and there should be a newline. Maybe add a reference to "goforit" somewhere too :)

File: flags.go [hunk 14]
```diff
@@ -84,6 +93,19 @@ func Enabled(ctx context.Context, name string) (enabled bool) {
...
 
+	defer func() {
+		flagsMtx.RLock()
+		staleMtx.RLock()
+		defer flagsMtx.RUnlock()
+		defer staleMtx.RUnlock()
+		staleness := time.Since(lastFlagRefreshTime)
+		//histogram of cache process age
+		stats.Histogram("goforit.flags.process_cache_age_s", staleness.Seconds(), nil, .01)
+		if staleness > staleness_threshold && time.Since(lastAssert) > 60*time.Second {
```
Comment (vasi-stripe) [inline flags.go]: Use a constant for the lastAssert interval (60 * time.Second)

File: flags.go [hunk 15]
```diff
@@ -19,14 +20,17 @@ const statsdAddress = "127.0.0.1:8200"
 
 var stats *statsd.Client
 
+var staleness_threshold time.Duration = 5 * time.Minute
```
Comment (vasi-stripe) [inline flags.go]: Go prefers camelCase to snake_case

File: flags.go [hunk 16]
```diff
@@ -187,16 +211,40 @@ func RefreshFlags(backend Backend) error {
 	for _, flag := range refreshedFlags {
 		fmap[flag.Name] = flag
 	}
-
+	if (age != time.Time{}) {
```
Comment (vasi-stripe) [inline flags.go]: You can us `age.isZero()` in this check

File: flags.go [hunk 17]
```diff
@@ -187,16 +211,40 @@ func RefreshFlags(backend Backend) error {
...
+		staleness := time.Since(age)
+		stale := staleness > staleness_threshold
+		//histogram of staleness
+		stats.Histogram("goforit.flags.cache_file_age_s", staleness.Seconds(), nil, .1)
+		//gauge
+		var gauge float64
+		if stale {
+			gauge = 1
+		}
+		stats.Gauge("goforit.flags.cache_file_stale", gauge, nil, 1)
```
Comment (vasi-stripe) [inline flags.go]: Ditch the gauge, we'll just log. Please add "goforit" somewhere in the log.

File: flags.go [hunk 18]
```diff
@@ -187,16 +211,40 @@ func RefreshFlags(backend Backend) error {
...
+		if(stale){
```
Comment (vasi-stripe) [inline flags.go]: Did you run `goimports`?

**Comment by vasi-stripe** [comment]
This is really close! Then we'll just need tests. I think the way to do that is probably to mock out the statsd client (which is sadly a global).

File: flags.go (hunk 12 above)
Comment (vasi-stripe) [reply flags.go]: Changed my mind, this one is ok cuz it's written to at very different times.

**Comment by vasi-stripe** [comment]
We paired, and came up with something that looks ok!

**Pull request merged**